enhancements/installer/aws-custom-edge-machineset-local-zones.md

```diff
@@ -0,0 +1,933 @@
+```
+
+You can find the raw files [here](https://github.com/mtulio/mtulio.labs/tree/master/labs/go-geo-app/data-benchmark).
+
+<!-- #### Communication Between Instances in different zone groups
+
+> TODO: check if there is any issue, it was mentioned by LZ experts. -->
+
+### Risks and Mitigations
+
+Administrators may want to add new locations on clusters already created by the instlaller.
+Phase-2 automates the creation of subnet(s) on Local Zones on Day-0, the CIDR blocks
+will be calculated automatically on the install time. If admins would like to expand
+they need to check if there's free CIDR to create subnets in the new locations, create
+the subnets and dependencies, then create the MachineSet manually.
+If there is no available CIDRs, they need to install a new cluster in existing
+subnets, and manage their own subnet CIDR blocks, creating the new locations (subnets
+and MachineSets) on Day-2.
+
+The total block allocated to the subnet on Local Zones could be underutilized,
+avoiding future expansions. This is a risk to the current IPI. A note on the documentation
+can be added: "if there is a plan to expand the subnets, the installation in existing VPC
+may be better".
+
+New Local Zones locations can be created without any type on the preferred list, then the MachineSet
+manifest generated will be created with an unsupported instance type, generating in `Failed`
+instance. We can keep the list of regions and locations tested on the Documentation, so we can
+provide better feedback on what is currently supported.
+
+### Drawbacks
+
+- Customers would prefer to deploy only on Day 2, or in existing VPC creating their own MachineSets
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Is there any situation to avoid the `NoSchedule` machine pools for edge workloads?
+
+Phase 2:
+
+1. The Local Zones currently have limitations of Nat Gateways, which keep the gateway
+  strategy more complex to provide egress traffic for compute nodes deployed on
+  that zone. Can we keep the edge nodes in the public subnet? Do we need to
+  create a dedicated security group instead of using/sharing with the default worker
+  machine pool?
+
+1. How we can keep extending the Zones for eventually new locations? AWS has a fixed
+  number of Local Zones available, when we create a cluster we will try to calculate
+  the subnet IPs to cover all IP ranges allowed by VPC (/16) depending on the number
+  of subnets available. I also wonder if we can consider a smaller range of IPs for
+  subnets operating in Local Zones, allowing the customer to grow it later. Can we
+  consider using a lower CIDR for subnets hosted in Local Zones locations?
+
+1. How we can advise customers to grow locations on Day 2? Can we keep a product
+  documentation page with that instructions? A sort of instructions on how to create
+  CloudFormation Stack to create a new zone for the new location that is available, or
+  the customer wants to support in the future.
```

Comment:
This is also likely to be useful for HyperShift, where... something? is creating subnets and such...  Although I'm not sure if that's on-demand at day-2, or something that happens at install-time and is assumed to persist throughout the cluster's lifetime.  It looks like [the NodePool's `subnet` property is an optional pointer][1], which hints at having some HyperShift-y logic for day-2 management.  But having Go installer code, instead of Terraform, that makes it easy to shift this into a day-2 controller at some point might be worthwhile.  Or maybe I'm overthinking this...

[1]: https://github.com/openshift/hypershift/blob/4073d66298251f2c7bb790c5be4afbec6af4dc28/api/v1beta1/nodepool_types.go#L569-L614

Reply:
I think expanding zones will be very common, mainly when we consider that Local Zones are quite new and are expanding around the globe continuously, and customers would like to use new locations in existing clusters but will be blocked as the installer consumed all the CIDRs blocks from the VPC when creating the subnets. I raised this question to the Arch call today, it's not clear if we can customize the CIDRs blocks of subnets on install time.

I don't have the data about *what users want*, but for me, it seems to be a very common day-2 operation, mainly when they want to create a smaller cluster and expand it continuously over time. I also think that we don't need to allocate big CIDR blocks for edge zones, for example, if the customer would like to be present in many edge zones within a given region just to deploy a few nodes there - why add a `/22`for a subnet which expects a few assets?! It's complex use cases, IMHO. Maybe it would be better to tell users to install in existing VPC and allow them to manage with their IPAM and expand in their end/automation, when the new location/zone should be added, they will create it on day 2 (network items and MachineSets).

Reply:
Section [`VPC CIDR block (Machine Network)`](https://github.com/openshift/enhancements/pull/1232/files#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R953-R995) added to the `Risks and Mitigations` detailing how to terraform will behave when deploying a cluster to prevent CIDR exhaustion, allowing subnet expansion on day 2.